sympy/functions/special/bessel.py

```diff
@@ -677,64 +677,82 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         except (ValueError, NotImplementedError):
             return self
 
-        if exp.is_positive and nu.is_integer:
-            newn = ceiling(n/exp)
-            bn = besseli(nu, z)
-            a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
-
-            b, c = [], []
-            o = Order(x**n, x)
+        # In case of powers less than 1, number of terms need to be computed
+        # separately to avoid repeated callings of _eval_nseries with wrong n
+        if exp.is_positive:
             r = (z/2)._eval_nseries(x, n, logx, cdir).removeO()
             if r is S.Zero:
-                return o
-            t = (_mexpand(r**2) + o).removeO()
+                return Order(z**(-nu) + z**nu, x)
 
-            if nu > S.Zero:
-                term = r**(-nu)*factorial(nu - 1)/2
-                b.append(term)
-                for k in range(1, nu):
-                    denom = (k - nu)*k
-                    if denom == S.Zero:
-                        term *= t/k
-                    else:
-                        term *= t/denom
-                    term = (_mexpand(term) + o).removeO()
-                    b.append(term)
+            o = Order(x**n, x)
+            if nu.is_integer:
+                # Reference: https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/01/04/01/02/0008/ (only for integer order)
+                newn = ceiling(n/exp)
+                bn = besseli(nu, z)
+                a = ((-1)**(nu - 1)*log(z/2)*bn)._eval_nseries(x, n, logx, cdir)
 
-            p = r**nu*(-1)**nu/(2*factorial(nu))
-            term = p*(digamma(nu + 1) - S.EulerGamma)
-            c.append(term)
-            for k in range(1, (newn + 1)//2):
-                p *= t/(k*(k + nu))
-                p = (_mexpand(p) + o).removeO()
-                term = p*(digamma(k + nu + 1) + digamma(k + 1))
+                b, c = [], []
+                t = (_mexpand(r**2) + o).removeO()
+
+                if nu > S.Zero:
+                    term = r**(-nu)*factorial(nu - 1)/2
+                    b.append(term)
+                    for k in range(1, nu):
+                        denom = (k - nu)*k
+                        if denom == S.Zero:
+                            term *= t/k
+                        else:
+                            term *= t/denom
+                        term = (_mexpand(term) + o).removeO()
+                        b.append(term)
+
+                p = r**nu*(-1)**nu/(2*factorial(nu))
+                term = p*(digamma(nu + 1) - S.EulerGamma)
                 c.append(term)
-            return a + Add(*b) + Add(*c) # Order term comes from a
+                for k in range(1, (newn + 1)//2):
+                    p *= t/(k*(k + nu))
+                    p = (_mexpand(p) + o).removeO()
+                    term = p*(digamma(k + nu + 1) + digamma(k + 1))
+                    c.append(term)
+                return a + Add(*b) + Add(*c) # Order term comes from a
+            elif nu.is_noninteger:
```

Comment:
This is not the same as `nu.is_integer is False` because `noninteger` implies `real`. Would it be acceptable here for `nu` to be a complex number like `1+I` or is it important that we only allow real `nu` here?

Reply:
While the formula in the given reference is correct (I think) for complex numbers, my implementation assumes nu is real. I have now added a comment mentioning this, and explicitly raised NotImplementedError for non-real order.

Reply:
How exactly does the implementation assume that `nu` is real?

Is it just the `ceiling((n+nu)/exp)` part?

Is it not easy to fix that so that it is valid for complex `nu`?

Ideally if there is a general expression that gives the series with all analytic coefficients then we would use it because it makes handling all of these different cases a lot simpler.

Reply:
![Wolfram general expression](https://github.com/user-attachments/assets/facf4421-8750-4a56-95e4-12dc87790ddd)

The above is the general expression from the Wolfram Mathworld link (implemented in the `nu.is_noninteger` branch). Note the $z^{\nu}$ and $z^{-\nu}$ in the two terms. For a power series expansion, we want the powers of $z$ to be real, and so directly putting complex $\nu$ into the series above is not useful. Nevertheless, the expression given above is still formally correct for complex order (just not useful as a power series expansion).

The trick used by Dunster (1990) is that when $\nu$ is purely imaginary (say $\nu = i \mu$, with $\mu$ being real), you can write $z^{i\mu} - z^{-i\mu} = 2 i \sin(\mu\log z)$, giving, after some manipulation, the series
![dunster 1990 eq 2.8](https://github.com/user-attachments/assets/e2c1296f-72bb-4507-aebe-93beca491198)

Thus, while the series I have used for the implementation of non-integer order is mathematically valid for non-real order as well, I don't see how I could implement the above handling of, e.g., purely imaginary order in `_eval_nseries` without making a new conditional branch.

Reply:
I've updated the comment 'for general complex order' -> 'for non-real order as well'